### PR TITLE
Improved mock generation with mocked network request responses

### PR DESCRIPTION
**Pull request by gurbirkalsi**
In order to support continued dashboard component development while the transition to the `pbench_server` is underway, the dashboard requires an advanced mock environment that allows for generation of response data and interception of network requests. By running the the dashboard with `yarn start:mock`, the mock development environment starts and listens for requests to intercept as defined in `mock/api.js`. This PR introduces a "fake data generator" library called casual to mock various response fields.

As server request postprocessing progress continues, the mock responses defined in `mock/api.js` will become significantly smaller. The primary motivation behind mocking requests is to allow for rapid prototyping of React components throughout the dashboard and other various feature development. For more details on the mock development environment, please reference the updated `README.md`.

### CODE REVIEW

File: src/utils/parse.test.js
Comment (gurbirkalsi) [inline src/utils/parse.test.js:23]: Please note we'll want to consider a new way for comparing mocked outcomes from our postprocessing without having the massive parsed response defined in our mock files. Eventually, this definition will move to the server as the server will handle postprocessing before sending a simplified response to the dashboard.

File: src/services/dashboard.js
Comment (gurbirkalsi) [inline src/services/dashboard.js:140]: As a first step, we'll define custom endpoints that have not yet been defined in `pbench_server`. Eventually, these will be replaced with their appropriate endpoints. These mocked endpoints are only available when the `MOCK_UI` env variable is set to True.

**Comment by anishaswain** [comment]
@gurbirkalsi I am experiencing an error such as
 ```
ExpandableTable.js: Uncaught TypeError: data.some is not a function
```
while trying to access dashboard with `yarn start:mock`. Any idea how to resolve the same?

**update**: _got resolved automatically with a few more try. Not sure how._

File: mock/api.js
Comment (aquibbaig) [inline mock/api.js]: We look for 
```
const index = endpoints.prefix + endpoints.run_index + indices[0];
```
in `search/fetchIndicexMapping` which according to the mock endpoints file becomes `test_prefix.v.run-data.2021-03` but the key given here is `test_index` which doesn't fetch mappings on to the search page.

**Review by aquibbaig — COMMENTED** [review]
Thanks for this PR @gurbirkalsi. It will be really helpful in running the dashboard in a mock environment.

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js]: Thanks for catching this Aquib! This is an interesting edge case where we're unable to build the correct index due to api.js being referenced as part of the Node build process which means we don't have access to `window.endpoints`. I'm including a temporary fix that allows us to reference `test_index` when in the mock development environment. We'll want to work on adapting the server endpoint for fetching index mappings to take an index parameter so that the dashboard doesn't parse the response for a specific index name.

**Comment by gurbirkalsi** [comment]
e2e tests are breaking due to the testing environment referencing our new mocked response generators. I will address the e2e tests once PR #124 is merged as most of the selectors for each page will change. Let me know if there are any other pending issues with this PR!

**Review by aquibbaig — APPROVED** [review]
LGTM

File: src/services/sessions.js
Comment (portante) [inline src/services/sessions.js:32]: Why not just usupr the `/graphql` endpoint for the mock'd environment as well?

**Review by portante — DISMISSED** [review]
Seems like we want to avoid knowledge of the mock'd environment as much as possible.  What can we do to avoid exposing that?

File: mock/api.js
Comment (dbutenhof) [inline mock/api.js]: Didn't Anisha have some outstanding code to _parse_ the mode string (once the effect actually returns it)? Is it worth building a better fake (octal numeric) mode now (or just hardcoding something like `"0o644"`?) vs deferring that to later?

File: mock/api.js
Comment (dbutenhof) [inline mock/api.js]: Similar comment here: should we worry about dashboard code that cares about the file type "keywords" returned from Elasticsearch and be more careful here about mocking? Maybe just hardcoded `"reg"` would be good enough?

File: mock/api.js
Comment (dbutenhof) [inline mock/api.js:251]: Not sure how much this'll matter for e2e testing, for for the record most of the properties are `"keyword"` now. For reference, here's an actual Postman response to the `_mappings` query:

```
{
  "drb.v6.run-data.2020-04": {
    "mappings": {
      "_meta": {
        "version": "6"
      },
      "date_detection": false,
      "properties": {
        "@generated-by": {
          "type": "keyword"
        },
        "@metadata": {
          "properties": {
            "controller_dir": {
              "type": "keyword"
            },
            "file-date": {
              "type": "date"
            },
            "file-name": {
              "type": "keyword"
            },
            "file-size": {
              "type": "long"
            },
            "md5": {
              "type": "keyword"
            },
            "pbench-agent-version": {
              "type": "keyword"
            },
            "raw_size": {
              "type": "long"
            },
            "result-prefix": {
              "type": "text"
            },
            "satellite": {
              "type": "keyword"
            },
            "tar-ball-creation-timestamp": {
              "type": "date"
            },
            "toc-prefix": {
              "type": "text"
            }
          }
        },
        "@timestamp": {
          "type": "date"
        },
        "authorization": {
          "properties": {
            "access": {
              "type": "keyword"
            },
            "owner": {
              "type": "keyword",
              "null_value": "NONE"
            }
          }
        },
        "host_tools_info": {
          "type": "nested",
          "properties": {
            "hostname": {
              "type": "keyword"
            },
            "hostname-f": {
              "type": "keyword"
            },
            "hostname-s": {
              "type": "keyword"
            },
            "label": {
              "type": "keyword"
            },
            "tools": {
              "properties": {
                "disk": {
                  "type": "text"
                },
                "haproxy-ocp": {
                  "type": "text"
                },
                "iostat": {
                  "type": "text"
                },
                "mpstat": {
                  "type": "text"
                },
                "oc": {
                  "type": "text"
                },
                "perf": {
                  "type": "text"
                },
                "pidstat": {
                  "type": "text"
                },
                "pprof": {
                  "type": "text"
                },
                "proc-interrupts": {
                  "type": "text"
                },
                "proc-vmstat": {
                  "type": "text"
                },
                "prometheus-metrics": {
                  "type": "text"
                },
                "sar": {
                  "type": "text"
                },
                "turbostat": {
                  "type": "text"
                },
                "vmstat": {
                  "type": "text"
                }
              }
            }
          }
        },
        "run": {
          "properties": {
            "config": {
              "type": "keyword"
            },
            "controller": {
              "type": "keyword"
            },
            "date": {
              "type": "date"
            },
            "end": {
              "type": "date"
            },
            "id": {
              "type": "keyword"
            },
            "iterations": {
              "type": "text"
            },
            "name": {
              "type": "keyword"
            },
            "script": {
              "type": "keyword"
            },
            "start": {
              "type": "date"
            },
            "toolsgroup": {
              "type": "keyword"
            },
            "user": {
              "type": "keyword"
            }
          }
        },
        "sosreports": {
          "type": "nested",
          "properties": {
            "hostname-f": {
              "type": "keyword"
            },
            "hostname-s": {
              "type": "keyword"
            },
            "inet": {
              "type": "nested",
              "properties": {
                "ifname": {
                  "type": "keyword"
                },
                "ipaddr": {
                  "type": "ip"
                }
              }
            },
            "inet6": {
              "type": "nested",
              "properties": {
                "ifname": {
                  "type": "keyword"
                },
                "ipaddr": {
                  "type": "keyword"
                }
              }
            },
            "md5": {
              "type": "keyword"
            },
            "name": {
              "type": "keyword"
            },
            "sosreport-error": {
              "type": "text"
            }
          }
        }
      }
    }
  }
}
```

File: mock/api.js
Comment (dbutenhof) [inline mock/api.js:307]: How are these URIs being interpreted? The server `/datasets/list` and `datasets/detail` APIs aren't in the server main yet; and while `/datasets/toc` and `/datasets/samples` are reasonable suggestions those server APIs aren't implemented; nor are native search and "sessions" (if such things will even exist in the current terms, which seems unlikely).

I guess I need to search to see where this is referenced, though; perhaps it'll make sense.

File: src/pages/document.ejs
Comment (dbutenhof) [inline src/pages/document.ejs:12]: FYI, I'm working on turning this into something like:

```
  <script type="text/javascript">
    fetch(`${window.origin}/api/v1/endpoints`)
      .then(function(response) {
        if (!response.ok) {
          console.log(`ERROR: ${response.statusText}: ${response.statusText}`);
        } else {
          response.json().then(function(data) {
            window.endpoints = data;
          })
        }
      });
  </script>
```

Although I'm not convinced that's the right packaging and I need to dump the `console.log` and invert that `if`. (Though this morning I've managed to break it every time I simplify anything.)

File: src/services/dashboard.js
Comment (dbutenhof) [inline src/services/dashboard.js:59]: Ah; interesting. Better to mock based on distinct endpoints rather than the klunky old JSON patterns from the payload; and eventually we can just simplify this to drop the elasticsearch alternative. I like.

File: src/services/search.js
Comment (dbutenhof) [inline src/services/search.js:40]: Wow; GitHub's rendition of this indentation change is really bizarre.

File: src/services/sessions.js
Comment (dbutenhof) [reply src/services/sessions.js:32]: Because the old mocks depended on JSON payload; the new ones are strictly URI endpoints, which is a lot cleaner and more reliable!

**Review by dbutenhof — APPROVED** [review]
This looks great. Some minor concern about the mocks of the file-level detail in the TOC, but I believe that doesn't impact the dashboard code until one of Anisha's outstanding PRs hits. (And currently that data, I believe, isn't even making it through the postprocessing effect.)

File: src/services/sessions.js
Comment (gurbirkalsi) [reply src/services/sessions.js:32]: Correct! We would be able to remove all references to the `MOCK_UI` once everything routes through the `pbench_server` endpoint. This is temporary solution in order to allow us to use the mock environment as soon as possible. Ideally, we can specify a single endpoint pointing to the `pbench_server` regardless of being in a mocked environment or not.

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js]: I'm certainly open to including a hardcoded value or creating a small helper to generate fake modes. @anishaswain Can you confirm if you're working on parsing the mode string as part of your work on table of contents data?

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js]: Thanks for pointing this out! Is there a definition for type of "keywords" we can expect from Elasticsearch? In that case, we can include a solution that randomly selects a keyword type from the list for each mocked object.

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js:251]: Interesting! I was referencing responses from the dashboard deployed in production which might be referencing an older version of Elasticsearch. For now, this doesn't cause problems because the dashboard is only concerned with seeing a string as part of the response raw value. It seems "keyword" and "text" equate to a string as part of the final raw value. Please correct me if I'm wrong here.

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js:307]: You'll notice in our `services/` definitions we specify `MOCK_UI` which is a runtime environment variable that allows us to determine if the dashboard is being run in a mocked environment. If that is the case, we tell the dashboard to hit these "fake" endpoints which ultimately return the mocked response. Over time, we can substitute these "fake" endpoints with real endpoints to `pbench_server` so we don't have to say `MOCK_UI ? /sessions/list : /graphl` when we build our endpoint url before submitting a request.

File: src/pages/document.ejs
Comment (gurbirkalsi) [reply src/pages/document.ejs:12]: Sweet! Looking forward to being able to query the endpoint configuration on demand.

File: mock/api.js
Comment (dbutenhof) [reply mock/api.js]: It's a lookup from the tar file type codes, in `indexer.py`:

```
    _mode_table = _dict_const(
        [
            (tarfile.REGTYPE, "reg"),
            (tarfile.AREGTYPE, "areg"),
            (tarfile.LNKTYPE, "lnk"),
            (tarfile.SYMTYPE, "sym"),
            (tarfile.DIRTYPE, "dir"),
            (tarfile.FIFOTYPE, "fifo"),
            (tarfile.CONTTYPE, "cont"),
            (tarfile.CHRTYPE, "chr"),
            (tarfile.BLKTYPE, "blk"),
            (tarfile.GNUTYPE_SPARSE, "spr"),
        ]
    )
```

File: mock/api.js
Comment (dbutenhof) [reply mock/api.js:251]: Yes, in ES1, we used `"string"` for most values. In ES7, indexed values are `"keyword"` and free-search text values (we don't have that many) are `"text"`.

Indeed, the dashboard mocking only needs to worry about what the dashboard cares about, and I don't believe it really cares about the type, only field names. Just thought I'd point this out... and aware that I posted way too much of the sample for that purpose. ;-)

File: src/pages/document.ejs
Comment (dbutenhof) [reply src/pages/document.ejs:12]: Actually, I was wondering whether you'd object philosophically to including that sort of code in the "root loader" html file. This is encouraging.

However, since you didn't object outright... this works beautifully on my VM, but I'm having problems in the e2e environment. The `fetch` is returning the full index.html, and the `.json()` is throwing because of the initial `<` character. (That's why I had the `console.log()` statements in here; although to actually nail this down I had to go to `response.text()` and print that before trying to manually jsonify it.) This may be partly because this is running too early, before the request mocking is set up, but getting the page html from the yarn server instead of a 404 still confuses me. Why would it ignore the path? Do you have any ideas here? Is there some yarn configuration, for example, that's making the development server return the page html for _every_ `GET` until we set up the mock overrides???

**Review by anishaswain — APPROVED** [review]
LGTM once the tests pass. Good work @gurbirkalsi

File: src/pages/document.ejs
Comment (gurbirkalsi) [reply src/pages/document.ejs:12]: My only concern is the added latency and overhead introduced in querying for a file before loading the dashboard bundle. Since the endpoint config is very small in size we can assume minimal latency and implement strategies for caching the definition in the future. 

Have you tried setting `MOCK=true` in the e2e environment? We currently do this in `package.json` for starting the dashboard in mocked environment: `"start:mock": "cross-env UMI_UI=none MOCK=true umi dev"`. It also may be helpful to debug the environment by setting the e2e tests to run without the headless mode: 

```
browser = await puppeteer.launch({
    headless: true -> false, // headless should be set to false 
    args: ['--no-sandbox'],
  });
```

Comment (dbutenhof) [reply src/pages/document.ejs:12]: In terms of HTTP protocol and Javascript, this is essentially identical to what you had before; a GET back to the origin, resulting in a JSON document, which is stashed in the `window` object for universal access. We do have the reverse proxy to the Pbench server instead of Apache's built-in file server, but we also have no disk access. I figure the performance and latency differences are going to be too small to measure.

Beyond that... it really seems like my problems are more a matter of how the npm dev server runs & is configured rather than in the Javascript itself. I tried copying your `start:mock` into my environment, but that doesn't appear to make any difference.

**Review by dbutenhof — APPROVED** [review]

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js]: I've updated the mock generator for the `type` field here to randomly select a tar file type upon generation.

File: mock/api.js
Comment (gurbirkalsi) [reply mock/api.js]: I've updated the `mode` field here to be a hardcoded value for "0o644".

**Comment by gurbirkalsi** [comment]
Merging the PR in its current state, e2e tests will be addressed with selector changes in PR #124

**Pull request merged**
